src/lib/ServoOutput/ServoMgr.cpp

```diff
@@ -15,38 +15,81 @@ ServoMgr::ServoMgr(const uint8_t *const pins, const uint8_t outputCnt, uint32_t
-uint8_t ServoMgr::allocateLedcChn(uint8_t ch, uint16_t intervalUs, uint8_t pin)
+#include "driver/ledc.h"
+
+#ifdef SOC_LEDC_SUPPORT_XTAL_CLOCK
+#define LEDC_DEFAULT_CLK LEDC_USE_XTAL_CLK
+#else
+#define LEDC_DEFAULT_CLK LEDC_AUTO_CLK
+#endif
+
+extern uint8_t channels_resolution[];
+
+/*
+ * Modified versions of the ledcSetup/ledcAttachPin from Arduino ESP32 Hal which allows
+ * control of the timer to use rather than being directly associated with the channel. This
+ * allows multiple channels to use the same timer within a LEDC group (high/low speed) when
+ * they are set to the same frequency.
+ */
+static void ledcSetupEx(uint8_t chan, ledc_timer_t timer, uint32_t freq, uint8_t bit_num)
+{
+    ledc_mode_t group = (ledc_mode_t)(chan / 8);
+
+    ledc_timer_config_t ledc_timer = {
+        .speed_mode = group,
+        .duty_resolution = (ledc_timer_bit_t)bit_num,
+        .timer_num = timer,
+        .freq_hz = freq,
+        .clk_cfg = LEDC_DEFAULT_CLK};
+    if (ledc_timer_config(&ledc_timer) != ESP_OK)
+    {
+        log_e("ledc setup failed!");
+        return;
+    }
+    channels_resolution[chan] = bit_num;
+}
+
+static void ledcAttachPinEx(uint8_t pin, uint8_t chan, ledc_timer_t timer)
+{
+    uint8_t group = (chan / 8), channel = (chan % 8);
+
+    ledc_channel_config_t ledc_channel = {
+        .gpio_num = pin,
+        .speed_mode = (ledc_mode_t)group,
+        .channel = (ledc_channel_t)channel,
+        .intr_type = LEDC_INTR_DISABLE,
+        .timer_sel = (ledc_timer_t)timer,
+        .duty = 0,
+        .hpoint = 0};
+    ledc_channel_config(&ledc_channel);
+}
+
+void ServoMgr::allocateLedcChn(uint8_t ch, uint16_t intervalUs, uint8_t pin)
 {
     uint32_t target_freq = 1000000U / intervalUs;
-    uint8_t ledc_chn = 255;
     _resolution_bits[ch] = (uint16_t)(log2f(80000000.0f / target_freq)); // no matter high speed timer or low speed timer, the clk src can be 80Mhz
     if (_resolution_bits[ch] > 16)
     {
         _resolution_bits[ch] = 16;
     }
 
-    for (int i = 7; i >= 0; --i)
+    uint8_t group = ch / 8;
+    for (int i = 0; i <= 3; i++)
     {
-        if (_timerConfigs[i].freq == 0)
+        uint8_t timer_idx = group * 4 + i;
+        if (_timerConfigs[timer_idx] == 0)
         {
-            _timerConfigs[i].freq = target_freq;
-            _timerConfigs[i].ch1 = ch;
-            ledc_chn = i * 2;
-            break;
+            _timerConfigs[timer_idx] = target_freq;
+            ledcSetupEx(ch, (ledc_timer_t)i, target_freq, _resolution_bits[ch]);
         }
-        else if (_timerConfigs[i].freq == target_freq && _timerConfigs[i].ch2 == 255)
+        if (_timerConfigs[timer_idx] == target_freq)
         {
-            _timerConfigs[i].ch2 = ch;
-            ledc_chn = i * 2 + 1;
-            break;
+            ledcAttachPinEx(pin, ch, (ledc_timer_t)i);
+            DBGLN("allocate ledc_ch %d on pin %d using group: %d, ledc_tim: %d, bits: %d", ch, pin, group, i, _resolution_bits[ch]);
+            return;
```

Comment:
here may be problems when 3 channels using the same freq?
like:
ch0 50hz
ch1 50hz
ch2 50hz

Reply:
No thats fine. All the channels in a group can share the same timer!

Reply:
> No thats fine. All the channels in a group can share the same timer!
Checked in the datasheet. You are right!